File: typescript/src/encoding.ts
/** USDC contract on Base mainnet (chain 8453). */
export const USDC_BASE = "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913";
export const USDC_DECIMALS = 6;

/** `transfer(address,uint256)` function selector. */
const ERC20_TRANSFER_SELECTOR = "0xa9059cbb";

/**
 * Builds the `data` field for an ERC-20 `transfer(to, amount)` call.
 * Returns hex string with 0x prefix.
 */
export function encodeErc20Transfer(to: string, amount: bigint): `0x${string}` {
  if (!/^0x[0-9a-fA-F]{40}$/.test(to)) {
    throw new Error(`Invalid address: ${to}`);
  }
  const addressPadded = to.slice(2).toLowerCase().padStart(64, "0");
  const amountPadded = amount.toString(16).padStart(64, "0");
  return `${ERC20_TRANSFER_SELECTOR}${addressPadded}${amountPadded}` as `0x${string}`;
}

/**
 * Converts a USD amount (decimal float) into USDC base units.
 * $0.10 → 100_000n (since USDC has 6 decimals).
 *
 * Uses string-based math to avoid IEEE-754 rounding on large values.
 */
export function usdToUsdcUnits(usd: number): bigint {
  if (!Number.isFinite(usd) || usd < 0) {
    throw new Error(`Invalid USD amount: ${usd}`);
  }
  const [whole, fraction = ""] = usd.toFixed(USDC_DECIMALS).split(".");
  const fractionPadded = fraction.padEnd(USDC_DECIMALS, "0").slice(0, USDC_DECIMALS);
  return BigInt(whole + fractionPadded);
}

/** Scans for entity-registry slugs. Anything not matching this is treated as a 0x address. */
export function isEntitySlug(to: string): boolean {
  return !/^0x[0-9a-fA-F]{40}$/.test(to);
}

File: python/canopy_ai/encoding.py
import re
from decimal import Decimal

# USDC contract on Base mainnet (chain 8453).
USDC_BASE = "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913"
USDC_DECIMALS = 6

_ERC20_TRANSFER_SELECTOR = "0xa9059cbb"
_ADDRESS_RE = re.compile(r"^0x[0-9a-fA-F]{40}$")


def encode_erc20_transfer(to: str, amount: int) -> str:
    """Builds the `data` field for an ERC-20 transfer(to, amount) call."""
    if not _ADDRESS_RE.match(to):
        raise ValueError(f"Invalid address: {to}")
    if amount < 0:
        raise ValueError(f"Amount must be non-negative: {amount}")
    address_padded = to[2:].lower().rjust(64, "0")
    amount_padded = f"{amount:x}".rjust(64, "0")
    return f"{_ERC20_TRANSFER_SELECTOR}{address_padded}{amount_padded}"


def usd_to_usdc_units(usd: float) -> int:
    """Converts a USD float to USDC base units (6 decimals)."""
    if usd < 0:
        raise ValueError(f"Invalid USD amount: {usd}")
    # Use Decimal to avoid float rounding errors.
    units = (Decimal(str(usd)) * (10**USDC_DECIMALS)).quantize(Decimal("1"))
    return int(units)


def is_entity_slug(to: str) -> bool:
    """Anything not matching the `0x…` address format is treated as an entity slug."""
    return _ADDRESS_RE.match(to) is None
